In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import Pipeline

In [2]:
data = pd.DataFrame({
    "Age": [20, 22, 25, 28, 30, 32, 35, 38, 40, 42, 45, 48, 50, 52, 55],
    "Salary": [25000, 28000, 30000, 35000, 40000, 45000, 50000, 55000,
               60000, 65000, 70000, 75000, 80000, 85000, 90000],
    "Experience": [1, 2, 2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 14, 16, 18],
    "Score": [55, 58, 60, 62, 65, 68, 70, 72, 75, 78, 80, 82, 85, 88, 90],
    "Target": [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1]
})

data

,Age,Salary,Experience,Score,Target
0,20,25000,1,55,0
1,22,28000,2,58,0
2,25,30000,2,60,0
3,28,35000,3,62,0
4,30,40000,4,65,0
5,32,45000,5,68,0
6,35,50000,6,70,0
7,38,55000,7,72,0
8,40,60000,8,75,0
9,42,65000,9,78,0


In [3]:
# Separate input features (X) and target variable (y)

X = data.drop("Target", axis=1)
y = data["Target"]

print("Features (X):")
print(X)

print("\nTarget (y):")
print(y)

Features (X):
    Age  Salary  Experience  Score
0    20   25000           1     55
1    22   28000           2     58
2    25   30000           2     60
3    28   35000           3     62
4    30   40000           4     65
5    32   45000           5     68
6    35   50000           6     70
7    38   55000           7     72
8    40   60000           8     75
9    42   65000           9     78
10   45   70000          10     80
11   48   75000          12     82
12   50   80000          14     85
13   52   85000          16     88
14   55   90000          18     90

Target (y):
0     0
1     0
2     0
3     0
4     0
5     0
6     0
7     0
8     0
9     0
10    0
11    0
12    1
13    1
14    1
Name: Target, dtype: int64


In [4]:
# Split the dataset into training and testing sets

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

print("\nTraining target distribution:")
print(y_train.value_counts())

print("\nTesting target distribution:")
print(y_test.value_counts())

X_train shape: (12, 4)
X_test shape: (3, 4)
y_train shape: (12,)
y_test shape: (3,)

Training target distribution:
Target
0    10
1     2
Name: count, dtype: int64

Testing target distribution:
Target
0    2
1    1
Name: count, dtype: int64


In [5]:
# Create a pipeline with StandardScaler and Logistic Regression

model = Pipeline([
    ("scaler", StandardScaler()),
    ("classifier", LogisticRegression(random_state=42))
])

# Train the model
model.fit(X_train, y_train)

print("Model trained successfully!")

Model trained successfully!


In [6]:
# Make predictions on the test set

y_pred = model.predict(X_test)

print("Actual values:    ", y_test.values)
print("Predicted values: ", y_pred)

print("Test Accuracy:", model.score(X_test, y_test))

Actual values:     [1 0 0]
Predicted values:  [1 0 0]
Test Accuracy: 1.0


In [7]:
# Evaluate the model using 5-fold cross-validation

cv_scores = cross_val_score(
    model,
    X,
    y,
    cv=5,
    scoring="accuracy"
)

print("Cross-validation scores:", cv_scores)
print("Mean CV accuracy:", cv_scores.mean())


Cross-validation scores: [1.         1.         0.66666667 1.         0.66666667]
Mean CV accuracy: 0.8666666666666666


c:\Users\ughan\CynarisInternship\AIML_PROJECT\venv\Lib\site-packages\sklearn\model_selection\_split.py:812: UserWarning: The least populated class in y has only 3 members, which is less than n_splits=5.
  warnings.warn(


## Cross-Validation Result

I evaluated the complete model pipeline using 5-fold cross-validation.

Cross-validation scores:
- Fold 1: 100%
- Fold 2: 100%
- Fold 3: 66.67%
- Fold 4: 100%
- Fold 5: 66.67%

Mean cross-validation accuracy: **86.67%**

### Observation

Cross-validation gives a more reliable estimate of model performance than relying only on the small test set.

A warning was displayed because the dataset contains only 3 samples from the minority class while 5 folds were used. This is a limitation of the small demonstration dataset. In a real project, a larger dataset would be used.

### Key Learning

The preprocessing step (`StandardScaler`) is included inside the Pipeline, so scaling is performed separately within each training fold. This helps prevent data leakage during cross-validation.